From 87854966cd2ed49d080dbbb8ca8f46101569f702 Mon Sep 17 00:00:00 2001
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Fri, 13 Apr 2018 05:51:27 -0300
Subject: [PATCH 1/2] vga: add share_surface flag

RH-Author: Gerd Hoffmann <kraxel@redhat.com>
Message-id: <20180413055129.17657-2-kraxel@redhat.com>
Patchwork-id: 79559
O-Subject: [RHEL-6.10 qemu-kvm PATCH 1/3] vga: add share_surface flag
Bugzilla: 1553674
RH-Acked-by: Stefan Hajnoczi <stefanha@redhat.com>
RH-Acked-by: Dr. David Alan Gilbert <dgilbert@redhat.com>
RH-Acked-by: Thomas Huth <thuth@redhat.com>

Add shares_surface bool variable, move the detection whenever we want
enable surface sharing a bit up in the code.  No functional change.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>
---
 hw/vga.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/hw/vga.c b/hw/vga.c
index f7de93b3e5..86302cfc2c 100644
--- a/hw/vga.c
+++ b/hw/vga.c
@@ -1792,6 +1792,7 @@ static void vga_draw_graphic(VGACommonState *s, int full_update)
     uint8_t *d;
     uint32_t v, addr1, addr;
     vga_draw_line_func *vga_draw_line;
+    bool share_surface = false;
 
     full_update |= update_basic_params(s);
 
@@ -1830,15 +1831,20 @@ static void vga_draw_graphic(VGACommonState *s, int full_update)
     }
 
     depth = s->get_bpp(s);
-    if (s->line_offset != s->last_line_offset ||
-        disp_width != s->last_width ||
-        height != s->last_height ||
-        s->last_depth != depth) {
 #if defined(HOST_WORDS_BIGENDIAN) == defined(TARGET_WORDS_BIGENDIAN)
-        if (depth == 16 || depth == 32) {
+    if (depth == 16 || depth == 32) {
 #else
-        if (depth == 32) {
+    if (depth == 32) {
 #endif
+        share_surface = true;
+    }
+
+    if (s->line_offset != s->last_line_offset ||
+        disp_width != s->last_width ||
+        height != s->last_height ||
+        s->last_depth != depth ||
+        share_surface != !!is_buffer_shared(s->ds->surface)) {
+        if (share_surface) {
             qemu_free_displaysurface(s->ds);
             s->ds->surface = qemu_create_displaysurface_from(disp_width, height, depth,
                     s->line_offset,
-- 
2.13.6

